### ISSUE TITLE
Provide a Flatpak as software package — for Linux systems

### ISSUE THREAD

**Issue by x80486** [standardnotes/desktop#172]
The title says it all!

Is there any plan to distribute **Standard Notes** as a Flatpak?

**Comment by mobitar** [comment]
Not familiar with Flatpak, but don't believe we'd do this ourselves as it creates more points of maintenance.

**Comment by mobitar** [comment]
We use electron-builder to distribute the app. So if they offer Flatpack at some point, I'd see no reason we wouldn't offer it as well :) https://github.com/electron-userland/electron-builder/issues/512

**Comment by x80486** [comment]
I may be wrong on this one.

My assumption (without even dig it) is that all Electron-based app are built pretty much the same way...if so, I think [this](https://github.com/atom/atom/pull/13293/files) can be a start.

Basically, what I'm trying to say is: if **Atom**, **Slack**, and **Standard Notes** are all Electron-based apps, and the first two are in Flathub already, then **Standard Notes** can be too 😇 ...of course, provided I know nothing about Electron 😉

**Comment by mobitar** [comment]
It's the autoupdate functionality that would be tricky. That's given to us by electron-builder. We'd have no solution when it comes to Flatpack, as far as I'm aware..

**Comment by x80486** [comment]
@mobitar, do you think you can make use of the [electron-installer-flatpak](https://www.npmjs.com/package/electron-installer-flatpak)?

I'm not entirely sure if it's _comparable_ with the method you are using to build **Standard Notes**, but maybe you can take a look at it whenever you have some free (what's that?!) time.

**Comment by mobitar** [comment]
Honestly I'm looking to make things even simpler going forward, in terms of what I have to manage and maintain. I've removed .deb support in the last release. Going to stick with AppImage for now because it supports autoupdates out of the box. If I do decide to switch package format in the future, I'd probably replace AppImage and not add more options.

**Comment by x80486** [comment]
OK, fair enough...since this is not a thing it makes sense to close it.

**Comment by x80486** [comment]
So, the `AppImage` stopped working again (I know why, but I wouldn't like to have more dangling dependencies because of that):

```sh
uplink org.gnome.Shell.desktop[1028]: #10 0x7ffefcc215c0 I   resource:///org/gnome/shell/ui/tweener.js:182 (0x7fd68c0cf560 @ 15)
uplink org.gnome.Nautilus[995]: /tmp/.mount_standahE6yRx/app/standard-notes: error while loading shared libraries: libgtk-x11-2.0.so.0: cannot open shared object file: No such file or directory
uplink gnome-shell[1028]: Object Meta.WindowActor (0x55678a9bb730), has been already finalized. Impossible to get any property from it.
```

Anyway, I went ahead and (try to) build a Flatpak :wink: ...but one of your dependencies (`"spellchecker": "github:mobitar/node-spellchecker"`) is failing in a strange way on the build phase:

```sh
  • electron-builder version=20.13.3
  • loaded configuration file=package.json ("build" field)
  • installing production dependencies platform=linux arch=x64 appDir=/run/build/standardnotes/main/app
Error: /app/bin/node exited with code 1
Error output:
npm info it worked if it ends with ok
npm info using npm@5.6.0
npm info using node@v8.11.3
npm info lifecycle standard-notes@2.3.4~preinstall: standard-notes@2.3.4
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/mobitar/node-spellchecker.git
npm ERR! 
npm ERR! ssh: Could not resolve hostname github.com: Temporary failure in name resolution
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128
...
Error: module standardnotes: Child process exited with code 1
```

Is it possible for you to reference this one in a different way? Like a standard `npm` package?

---

I can access GitHub via SSH. Right after that failure, I can execute this successfully:

```
$ ssh -vT git@github.com
OpenSSH_7.7p1, OpenSSL 1.1.0i  14 Aug 2018
debug1: Reading configuration data /home/x80486/.ssh/config
debug1: /home/x80486/.ssh/config line 2: Applying options for github.com
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Connecting to github.com [192.30.253.113] port 22.
debug1: Connection established.
debug1: identity file /home/x80486/.ssh/github_rsa type 0
...
debug1: Authenticating to github.com:22 as 'git'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
...
debug1: Host 'github.com' is known and matches the RSA host key.
debug1: Found key in /home/x80486/.ssh/known_hosts:1
...
debug1: Offering public key: RSA SHA256:e...Q /home/x80486/.ssh/github_rsa
debug1: Server accepts key: pkalg ssh-rsa blen 535
debug1: Authentication succeeded (publickey).
Authenticated to github.com ([192.30.253.113]:22).
...
Hi x80486! You've successfully authenticated, but GitHub does not provide shell access.
debug1: channel 0: free: client-session, nchannels 1
Transferred: sent 3360, received 2024 bytes, in 0.1 seconds
Bytes per second: sent 57386.0, received 34568.2
debug1: Exit status 1
```

**Comment by mobitar** [comment]
Hmm..it may be that the submodule is indeed added through ssh. Can you see if this command works for you:

```
git clone git@github.com:mobitar/node-spellchecker.git
```

I probably should have added it through https for simplicity. Can try to make this update in the next release.

**Comment by x80486** [comment]
Yes, I can actually do anything, but so far, that's the only thing I can tell that it's failing:

```sh
[x80486@uplink no_backup]$ git clone git@github.com:mobitar/node-spellchecker.git
Cloning into 'node-spellchecker'...
remote: Counting objects: 1142, done.
remote: Total 1142 (delta 0), reused 0 (delta 0), pack-reused 1142
Receiving objects: 100% (1142/1142), 11.37 MiB | 29.40 MiB/s, done.
Resolving deltas: 100% (593/593), done.
```

Maybe it's a limitation of the build tool (on the Flatpak side) – I know there is a step to bring over all dependencies offline for repeateable builds and what not.

If you find a way to change it, I'll give it a try again, otherwise I'll wait for the whole build process to get better I guess.

**Comment by mobitar** [comment]
Ok, have made a note to get around to this. You can also edit this yourself in the meantime by editing the .gitmodules file and the .git/config file to change from ssh to https.

**Comment by x80486** [comment]
Nevermind. Don't waste your time on this one. The dependencies are downloaded in order to build the artifact offline. That's how the build process works, so it will never reach that dependency.

I'll see what I can do :no_good_man:

**Comment by x80486** [comment]
I'll reopen this one and also leave [this work](https://github.com/x80486/flathub/tree/org.standardnotes.standardnotes/) in case a more knowledgeable soul in `npm`, `Electron` (and alike) wants to give it a whirl. The build process fails while _building_ `mobitar/node-spellchecker`, but my guess is that this is due to a bad marriage between `npm` / `node-gyp` / Flatpak. I can't tell so far.

**Comment by mobitar** [comment]
Hmm..do you have console output for the error?

**Comment by mobitar** [comment]
Oh sorry, just scrolled up. So this is that same ssh issue? I'll update this to use https hopefully this week.

**Comment by x80486** [comment]
No that's fine the way it is. I just get passed that. It's now failing trying to build `mobitar/node-spellchecker` on this operation: `prebuild-install || node-gyp rebuild`. I'll update this one later on with the message itself.

**Comment by x80486** [comment]
Yeah, but I would need to pull those from a fixed location, not build them inside the build process itself.

One example:

```json
{
  "type": "file",
  "only-arches": ["i386"],
  "sha256": "0ba392aca71d85cce5465e1431eb6c35c8f0609e598c485cc67177a99827bc99",
  "url": "https://github.com/wireapp/wire-desktop/releases/download/release/3.3.2872/wire_3.3.2872_i386.deb"
},
{
  "type": "file",
  "only-arches": ["x86_64"],
  "sha256": "fca2cabdd934dca956525b60f97dd2712120499c7f10b10259fe4701439ada13",
  "url": "https://github.com/wireapp/wire-desktop/releases/download/release/3.3.2872/wire_3.3.2872_amd64.deb"
},
{
  "type": "file",
  "sha256": "35949658604dbd400f3d86925b5ef3754c1ead285324dd94d2d0ed1e2d485bc4",
  "url": "https://github.com/wireapp/wire-desktop/archive/release/3.3.2872.tar.gz"
},
```

**Comment by r3l4t1v** [comment]
@mobitar would you consider merging a PR that adds support for building a flatpak?

**Comment by x80486** [comment]
Hey folks, I gave it another whirl to this since the `spellchecker` dependency is not there anymore...but now `standard-notes-web` (in `https://github.com/standardnotes/desktop/blob/dev/app/package.json`) is the issue.

I was wondering if you can declare that dependency in a different way (like the others), something like `"standard-notes-web": "3.3.1"`? I see that dependency already in NPM: https://www.npmjs.com/package/standard-notes-web/

**Comment by x80486** [comment]
All right! I was able to build this from sources! :partying_face:

There are some issues that I think, if you are interested in moving forward with this format, should be fixed first before publishing it, because Flatpak is sandboxed by default, hence, it wouldn't make sense to expose permissions that shouldn't be set.

Here is the list of issues:

- The application should follow the Freedesktop XDG base directory standard, even for backups (should be extremely easily to change)

- Disable the update manager because nothing can be written inside the container. There is a file: `/.flatpak-info`, if it exists, don't do the update check; I see this error all the time I start it:

```shell
updateManager: Checking for updates.
14:54:16.548 › Checking for update
14:54:16.571 › Error: Error: ENOENT: no such file or directory, open '/app/standardnotes/resources/app-update.yml'
updateManager: Exception caught while checking for autoupdates: [Error: ENOENT: no such file or directory, open '/app/standardnotes/resources/app-update.yml'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/app/standardnotes/resources/app-update.yml'
}
updateManager: Finished checking for updates. Latest version:
 3.3.5 Current version: 3.3.5
```

Nice to have:

- Modify or provide another build step for `dist-linux`
- Use Yarn instead of `npm` (this is actually [recommended](https://www.electron.build/#installation) for Electron applications that use `electron-builder`)
- Provide an SVG icon

**Comment by baptiste-grob** [comment]
That's really cool, great job! I have two followup questions:

- Where should backups be if we were to follow the Freedesktop XDG base directory standard? something like `~/.standardnotes`?
- If we disable the auto-updater, I assume we'd have to publish the flatpak somewhere. Is there a popular flatpak store/repo?

**Comment by x80486** [comment]
My understanding is that, for Linux, it should be `${XDG_DATA_HOME}` (which translates to something like: "data files" go into `${HOME}/.local/share/`), hence: `${XDG_DATA_HOME}/standardnotes/`. I've [read something](https://www.electronjs.org/docs/api/app#appgetpathname) about Electron and it looks like `app.getPath(name)` is the way to go, providing the appropriate `name` identifier, because it will give you the equivalent for all platforms (but you probably know that already). Again, I think that should be done regardless of the Flatpak, but the reason I proposed it in the Flatpak context is because you wouldn't need to widen the permissions due to a design flaw; of course, you can release it like that, and tweak it in the future (IntelliJ IDEA just fixed that).

Let me be clear, the auto-updater should only be disabled for the Flatpak, any other format(s) can make use of it. Being said that, [Flathub](https://flathub.org/) is the store that I know everyone uses today – think of it like the Play Store (for Android) or the Apple Store. They take care about the entire infrastructure for building, publishing, etc.

**Comment by baptiste-grob** [comment]
Gotcha, we don't have much bandwidth to add another publishing format at the moment, but it's good to know for future reference!
We're already using `app.getPath(...)` for backups. But we give it the `home` identifier, because we want backups to be easily discoverable by the user. They're here in case of a major problem occurring after all, so burying them alongside technical application files is not desirable

**Comment by x80486** [comment]
OK, it's really your decision as the owner of Standard Notes.

Now coming back to the _bandwidth_ point. The reason I did it was because this is much better than the AppImage (which is broken in many ways some times, at least for me). So I'm willing to maintain the format for some time (which I would say, it's a matter of updating a couple of lines in JSON file), but it would be better if you (and/or some others from Standard Notes) are part of the (GitHub) repo in Flathub, at some point, if it turns out that Flatpak works great for this application to promote it officially, and things like that...my point is: this could be driven by us (the community) but it would be better if it's supported and not an orphan thing from your side.

Let me know.

**Comment by makruiten** [comment]
I think Flatpak the way forward. Flatpak is already present on many Linux distributions by default and is very well integrated into the system. Flathub is a very popular Flatpak repository and is also included in many distributions (either enabled, or a few clicks away from being enabled). Applications can be easily discovered and (automatically) updated from the software centre (GNOME Software for example).

**Comment by baptiste-grob** [comment]
To be clear, we will absolutely look at supporting Flatpak in the future, including how much we'd involve the community in that process 🙂 in the meantime we'll keep supporting Snaps and AppImage.

**Comment by A6GibKm** [comment]
@x80486  could you please share the flatpak manifest to try building it myself?

> The application should follow the Freedesktop XDG base directory standard 

Note that, while following xdg standard is a good practice, it is not required for this app as one could simply build with the flags `--filesystem=~/Standard\ Notes\ Backups` if we want it to access the folder in the user home, or `--persist=Standard\ Notes\ Backups` for it to be stored inside the sandbox (at `~/.var/app/org.standardnotes.standardnotes/Standard\ Notes\ Backups`). User could give it `--filesystem=home` if they don't want to overthink it when moving the location of the backups.

> Disable the update manager because nothing can be written inside the container
Could this crash the app?

I would be happy to keep an updated build at flathub, (bundle icons, desktop file, provide an appdata.xml if needed, and the publish process), so that the devs can take over it when/if they feel comfortable with this distribution method. But for that, having an already working manifest would help a lot.

**Comment by x80486** [comment]
Hey @A6GibKm, I don't have it anymore. If they decide to move forward, I'll put it together again (it's super-easy now that I know how to do it :rofl: ) and handle it to you.

Pay attention to the thread :point_up:, even if you have the manifest right now, they won't agreed on publishing it in Flathub, and as soon as you create a pull request over there they are going to ask you if the developers are OK with the app on their hub.

**Comment by justyn** [comment]
How about a temporary, unofficial Standard Notes Flatpak called something else? Perhaps "Ordinary Notes".

A community maintained "friendly fork" with no other changes, that would enable those who want to use Flatpak to do so, coming with no guarantees of working whatsoever? It could be folded into an official Standard Notes Flatpak in the future if/when the SN devs so wish. In the meantime it could be a useful experiment for them to keep an eye on, without taking up any of their time.

This is a well-intentioned suggestion so apologies if it inadvertently causes any offence! I don't have the time to dedicate to such a project myself anyway, but I note that a few people here seem to be keen to put the effort in!

**Comment by rstrube** [comment]
As a Standard Notes subscriber, I'd also love to see official Flatpak support. It's great that x80486 was able to get something up and running!.  Lots of folks aren't very pleased with Snaps for a variety of reasons including a proprietary backend, and forced updates.  Here's an example of some of the frustrations a fellow developer has delt with in regards to snaps: https://raymii.org/s/blog/Ubuntu_Snap_auto_updates_broke_my_development_setup.html.  Also distros like Mint and Pop_OS have officially removed snapd and embraced Flatpak.

Please reconsider adding official flatpak support, many of your Linux customers would really appreciate it!

**Comment by baptiste-grob** [comment]
> To be clear, we will absolutely look at supporting Flatpak in the future, including how much we'd involve the community in that process 🙂 in the meantime we'll keep supporting Snaps and AppImage.

This comment is still valid, we're slammed with other work that can't be delayed so we're just not in that future yet! Hope you'll bear with us and thanks for subscribing. :)

**Comment by baarkerlounger** [comment]
@baptiste-grob worth re-opening this if it's something the StandardNotes team are still planning to look at in the future?

**Comment by ktleary** [comment]
Adding another vote for flatpak, as due to my own fault (hardening firefox beyond what is reasonable for this type of app) I get wasm errors (e.g. "Both wasm and asm failed to load", "The Same Origin Policy disallows reading the remote resource at https://notify.bugsnag.com",   ...). A flatpak distribution would be a great.

**Comment by baptiste-grob** [comment]
Good point @baarkerlounger :)

**Comment by rendhalver** [comment]
I would also like a flatpak version of Standard-Notes.
I would be interested in testing things and maybe helping out with the development and release automation if that helps.

**Comment by jadmadi** [comment]
I would also support the notion of having Standard Notes on Flatpaks hub

**Comment by icfmp** [comment]
Yes please, provide it on flathub :+1:

**Comment by dtantono** [comment]
There is an unofficial flatpak (https://flathub.org/apps/details/org.standardnotes.standardnotes). But it also asks for keyring every time I reboot my PC. that means I need to enable password reading on the startup-like snap app which is quite annoying

**Comment by vchernin** [comment]
There is now experimental support for building flatpaks in electron-builder. There is no native support to upload directly to flathub yet. See https://github.com/electron-userland/electron-builder/pull/5711
